cpp/CONTRIBUTING.md

```diff
@@ -0,0 +1,184 @@
+output block when the recipe is rendered into the cookbook.
+
+Referencing Arrow C++ Documentation
++++++++++++++++++++++++++++++++++++
+
+The Arrow project has its own documentation for the C++ implementation that
+is hosted at https://arrow.apache.org/docs/cpp/index.html.  Fortunately,
+this documentation is also built with Sphinx and so we can use the extension
+`intersphinx` to reference sections of this documentation.  To do so simply
+write a standard Sphinx reference like so:
+
+```
+Typed subclasses of :cpp:class:`arrow::ArrayBuilder` make it easy
+to efficiently create Arrow arrays from existing C++ data
+```
+
+A helpful command is
+`python -msphinx.ext.intersphinx https://arrow.apache.org/docs/objects.inv`
+which will list all of the possible targets to link to.
+
+Development Workflow
+====================
+
+Running `make` at the top level can be rather slow as it will rebuild the
+entire environment each time.  It is primarily intended for use in CI and
+requires you to have conda installed.
+
+For recipe development you are encouraged to create your own out-of-source
+cmake build.  For example:
+
+```
+mkdir cpp/code/build
+cd cpp/code/build
+cmake ../code -DCMAKE_BUILD_TYPE=Debug
+cmake --build .
+ctest
+```
+
+Then you can rerun all of the tests with `ctest` and you can rebuild and
+rerun individual tests much more quickly with something like
+`cmake --build . --target creating_arrow_objects && ctest creating_arrow_objects`.
+Everytime the cmake build is run it will update the recipe output file
+referenced by the sphinx build so after rerunning a test you can visualize the
+output by running `make html` in the `cpp` directory.
+
+Using Conda
++++++++++++
+
+If you are using conda then there is file `cpp/requirements.yml` which can be
+used to create an environment for recipe development with the command:
+
+```
+conda env create -n cookbook-cpp --file cpp/requirements.yml
+```
+
+Development Philosophy
+======================
+
+Everything is the Cookbook
+++++++++++++++++++++++++++
+
+The entire document should serve as an example of how to use Arrow C++, not just the
+referenced snippets.  This means that the below style rules and guidelines apply to
+source code that is not referenced by the cookbook itself.
+
+Style
++++++
+
+This cookbook follows the same style rules as Arrow C++ which is the Google style
+guide with a few exceptions described
+[here](https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci)
+
+Simple
+++++++
+
+The examples should be as simple as possible.  If complex code (e.g. templates) can be
+used to do something more efficiently then there should be a simple, inefficient version
+alongside the more complex version.
+
+Do not use `auto` in any of the templates unless you must (e.g. lambdas).  Cookbook
+viewers will be using a browser, not an IDE, and it is not always simple to determine
+the inferred type.
+
+For conciseness, `using namespace std;` should be used.
```

Comment:
I'm not sure the conciseness buys us much compared with potential name clashes/confusion over where symbols come from. Also, I don't see this rule in the Arrow source and in fact the only instance of this declaration is in vendored third-party code.

Reply:
It also explicitly contradicts the C++ style guide: https://google.github.io/styleguide/cppguide.html#Namespaces

Reply:
Yes, sorry, I didn't mean to imply it was part of the Arrow style guide.  I was trying to call it out as an exclusion.  The `std` namespace *should* be pretty well known and, if it is the only one we do a `using` for then it shouldn't be ambiguous (the examples don't live in a namespace).

But...it's also only 5 characters (although the default rendering only gives us 72) so it's not that big a deal.

Reply:
Ah, ok. Yeah, I would feel it's best not to deviate from Arrow itself on this one. I feel like aliases for Arrow namespaces would probably save more space (ds = ::arrow::dataset and so on).

Reply:
Ok, I've removed `using std`.  At the moment things are ok.  I like the idea of arrow namespace aliases.  I think we can just play it by ear for now.